shared/middleware/helpers/file_bucket.rb

```diff
@@ -13,14 +13,37 @@ def allowed_file_type?(extension)
     Sinatra::Base.mime_type(extension)
   end
 
+  def get_manifest(channel_id)
+    manifest_result = get(channel_id, FileBucket::MANIFEST_FILENAME)
+    if manifest_result[:status] == 'NOT_FOUND'
+      return []
+    end
+    JSON.load manifest_result[:body]
+  end
+
   def copy_files(src_channel, dest_channel, options={})
     # copy everything except the manifest
     options[:exclude_filenames] = [MANIFEST_FILENAME]
     result = super src_channel, dest_channel, options
+    # return right away if there are no files in the source project
+    return if result.empty?
+
+    # update dest_manifest with the decorated filenames from the src_manifest
+    # (we need the new version ids from the dest_manifest, otherwise we'd just copy the src_manifest)
+    dest_manifest = JSON.load result.to_json
+    src_manifest = get_manifest(src_channel)
+    src_manifest.each do |src_entry|
+      src_filename = src_entry['filename']
+      dest_manifest.each do |dest_entry|
+        if dest_entry['filename'].downcase == src_filename.downcase
+          dest_entry['filename'] = src_filename
+        end
+      end
+    end
```

Comment:
While I don't think we're enforcing a limit on the size of n here, it should be < 10 in the vast majority of cases, so yeah - I'm premature-optimizing.  I tend to do that with server code.  Also, FilesAPI is already the source of many of our slowest transactions (says New Relic) and while the S3 round-trip is clearly most of that cost, I'm still sensitive to it.

I'll make an argument that my version is a little more readable too because it separates concerns and removes a level of nesting.  The first five lines could be easily extracted to a helper method.